lib/lua/checkStalledJobs.lua

```diff
@@ -13,30 +12,38 @@ if a jobId is not removed from the stalling set within a stallInterval window,
 we assume the job has stalled and should be reset (moved from active back to waiting)
 --]]
 
-local now = tonumber(ARGV[1])
-local stallTime = tonumber(redis.call("get", KEYS[1]) or 0)
-
-if now < stallTime then
+-- try to update the stallBlock key
+if not redis.call("set", KEYS[1], "1", "PX", tonumber(ARGV[1]), "NX") then
   -- hasn't been long enough (stallInterval) since last check
-  return 0
+  return {}
 end
 
 -- reset any stalling jobs by moving from active to waiting
-local stalling = redis.call("smembers", KEYS[2])
-if #stalling > 0 then
-  redis.call("rpush", KEYS[3], unpack(stalling))
-  for i = 1, #stalling do
-    redis.call("lrem", KEYS[4], 0, stalling[i])
+local stalling, stalled = redis.call("smembers", KEYS[2]), {}
+if next(stalling) ~= nil then
+  -- not worth optimizing - this should be a rare occurrence, better to keep it straightforward
+  local nextIndex = 1
+  for i, jobId in ipairs(stalling) do
+    local removed = redis.call("lrem", KEYS[4], 0, jobId)
+    -- safety belts: we only restart stalled jobs if we can find them in the active list
+    -- the only place we add jobs to the stalling set is in this script, and the two places we
+    -- remove jobs from the active list are in this script, and in the MULTI after the job finishes
+    if removed > 0 then
+      stalled[nextIndex] = jobId
+      nextIndex = nextIndex + 1
+    end
+  end
+  -- don't lpush zero jobs (the redis command will fail)
+  if nextIndex > 1 then
+    redis.call("lpush", KEYS[3], unpack(stalled))
```

Comment:
This is just to avoid stalled jobs hogging the consumers if for some reason a specific job causes an uncaught exception. I'll add an explanatory comment.